mender-image-prepare.sh

```diff
@@ -0,0 +1,535 @@
+#!/bin/bash
+
+show_help() {
+cat << EOF
+Usage: $0 [prepare_image | make_sdimg | make_sdcard] [-i|--image | -d|--device -r|--rootsize]
+
+        device - can be /dev/sd* or /dev/mmcblk*
+
+        prepare_image - shrinks existing embedded image
+                        obligatory parameters: --image
+
+        make_sdcard   - formats, partitions and sets the sd card
+                        obligatory parameters: --image & --rootsize
+                        
+                        note: rootsize can found as an output from
+                        'prepare_image' command
+
+        make_sdimg    - composes image file compliant with Mender partition
+                        table scheme
+
+EOF
+}
+
+if [ $# -eq 0 ]; then
+  show_help
+  exit 1
+fi
+
+if ! id | grep -q root; then
+        echo "must be run as root"
+        exit
+fi
+
+BASE_DIR=${PWD}
+
+# Erase block size 8MiB
+erase_block=8388608
+# Boot partition start in sectors (512 bytes per sector).
+pboot_start=16384
+# Boot partition size in sectors.
+pboot_size=$(( 16777216 / 512 ))
+# Data partition size in sectors.
+pdata_size=$(( 134217728 / 512 ))
+
+# beaglebone: 4521984
+prootfs_size=
+target_dev=
+image=
+number_of_parts=
+loopsdimg=
+
+declare -a part_names=("boot" "primary" "secondary" "data")
+declare -a mappings
+
+get_part_number_from_device() {
+    case "$1" in
+        /dev/*[0-9]p[1-9])
+            echo ${1##*[0-9]p}
+            ;;
+        /dev/[sh]d[a-z][1-9])
+            echo ${1##*d[a-z]}
+            ;;
+        ubi[0-9]_[0-9])
+            echo ${1##*[0-9]_}
+            ;;
+        [a-z]*\.sdimg[1-9])
+            echo ${1##*\.sdimg}
+            ;;
+        /dev/mapper/*[0-9]p[1-9])
+            echo ${1##*[0-9]p}
+            ;;
+        *)
+            echo "Could not determine partition number from $1"
+            exit 1
+            ;;
+    esac
+}
+
+unmount_partitions() {
+  echo "1. Check if device is mounted..."
+  is_mounted=`grep ${target_dev} /proc/self/mounts | wc -l`
+  if [ ${is_mounted} -ne 0 ]; then
+    umount ${target_dev}?*
+  fi
+}
+
+erase_part_table() {
+  echo "2. Erase partition table..."
+  wipefs --all --force ${target_dev}?*
+  dd if=/dev/zero of=${target_dev} bs=1M count=100
```

Comment:
This erases quite a lot more than just the partition table. Perhaps it should be called `erase_filesystem` instead?

Reply:
Ok, I will rename the method name.

Reply:
I kinda expected for this script only to output a `.sdimg` and not actually write it to a physical device. A bit more modular approach and the script is called "mender-image-prepare", we could have a separate script that is "mender-image-write" if we want.

Thoughts?

Reply:
Ok, thing to think through.